### PR TITLE
FNO nn.Module class with positional encoding using meshgrids

### PR DESCRIPTION
**Pull request by spellsharp**
## Description
Introduces a base Fourier Neural Operator (FNO) as a PyTorch nn.Module for DeepChem, with unit tests and docs wiring. The module implements the standard FNO stack (lifting → FNO blocks with spectral + local ops → projection) and accepts channel-first or channel-last inputs via an internal shape normalizer. The PR also exposes FNO in `deepchem.models.torch_models.__init__` [GitHub](https://github.com/deepchem/deepchem/pull/4491/files)

What’s new
- New FNO class in deepchem/models/torch_models/fno.py:
- Configurable in_channels, out_channels, modes, width, dims, depth.
- Lifting MLP, a nn.Sequential of FNOBlocks, and a projection head.
- _ensure_channel_first to accept (B, C, *spatial) or (B, *spatial, C). [GitHub](https://github.com/deepchem/deepchem/pull/4491/files)
- Tests for the FNO module (tests/models/torch_models/test_fno.py). [GitHub](https://github.com/deepchem/deepchem/pull/4491/files)
- Docs/API surface updated to list the new class; FNO exported from __init__.py.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/utils/fno_utils.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import torch
+from typing import Optional, List
+
+
+class GaussianNormalizer:
```
Comment (JoseAntonioSiguenza) [inline deepchem/utils/fno_utils.py]: please add unit tests for utils

File: deepchem/models/torch_models/fno.py [hunk 2]
```diff
@@ -83,3 +84,181 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         x1 = self.spectral_conv(x)
         x2 = self.w(x)
         return F.relu(x1 + x2)
+
+
+class FNO(nn.Module):
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/fno.py:88]: Can you please add this class to the docs?

File: deepchem/models/torch_models/fno.py [hunk 3]
```diff
@@ -83,3 +84,181 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+                 modes: Union[int, Tuple[int, ...]],
+                 width: int,
+                 dims: int,
+                 depth: int = 4,
+                 positional_encoding: bool = False,
+                 normalize_input: bool = True,
+                 normalize_output: bool = True,
+                 normalization_dims: Optional[List[int]] = None) -> None:
+        """Initialize the FNO base model.
+        Parameters
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/fno.py:124]: Please move `Parameters` section to `FNO` class docstring

File: deepchem/models/torch_models/fno.py [hunk 4]
```diff
@@ -83,3 +84,181 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        self.lifting = nn.Sequential(nn.Linear(self.in_channels, 2 * width),
+                                     nn.GELU(), nn.Linear(2 * width, width))
+
+        self.fno_blocks = nn.Sequential(
+            *[FNOBlock(width, modes, dims=dims) for _ in range(depth)])
+
+        self.projection = nn.Sequential(nn.Linear(width, 2 * width), nn.GELU(),
+                                        nn.Linear(2 * width, out_channels))
+
+    def fit_normalizers(self,
```
Comment (JoseAntonioSiguenza) [inline deepchem/models/torch_models/fno.py]: Please improve the docstrings of all methods. What do they return? Also, please describe inputs and include usage examples.

File: deepchem/utils/fno_utils.py (hunk 1 above)
Comment (JoseAntonioSiguenza) [inline deepchem/utils/fno_utils.py]: Please add usage examples for methods and add more explanation on `GaussianNormalizer`. Add references/citations if you are adapting this code from an exisiting implementation.

Comment (spellsharp) [reply deepchem/utils/fno_utils.py]: Added usage example. 

>Please add usage examples for methods and add more explanation on GaussianNormalizer. Add references/citations if you are adapting this code from an exisiting implementation. 

I'm not particularly referring to any existing implementation. It is just math in pytorch.

File: deepchem/models/torch_models/fno.py (hunk 4 above)
Comment (spellsharp) [reply deepchem/models/torch_models/fno.py]: Done

File: deepchem/models/torch_models/fno.py (hunk 2 above)
Comment (spellsharp) [reply deepchem/models/torch_models/fno.py:88]: Done

File: deepchem/utils/fno_utils.py (hunk 1 above)
Comment (spellsharp) [reply deepchem/utils/fno_utils.py]: Done

File: deepchem/models/torch_models/fno.py (hunk 3 above)
Comment (spellsharp) [reply deepchem/models/torch_models/fno.py:124]: I'm sorry for the confusion. In most of the torch_model files I looked at, the class docstring only has a class description, references, and a usage example. The parameters are mostly found in the __init__ method. Should I still add parameters to class docstring?

File: deepchem/utils/fno_utils.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
+import torch
+from typing import Optional, List
+
+
```
Comment (rbharath) [inline deepchem/utils/fno_utils.py]: How does this compare with deepchem transformers' NormalizationTransformer class? When would we use this instead of that?

**Comment by JoseAntonioSiguenza** [comment]
LGTM

**Review by rbharath — COMMENTED** [review]
LGTM once CI is back online for unit tests

**Review by rbharath — APPROVED** [review]
LGTM
